In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))
        


# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/titanic/train.csv
/kaggle/input/titanic/test.csv
/kaggle/input/titanic/gender_submission.csv
/kaggle/input/d/hsingh956/titanic/train.csv
/kaggle/input/d/hsingh956/titanic/test.csv


# # **Importing the libraries**


In [2]:
import numpy as np
import pandas as pd
import os
# # # # pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 30)
# # pd.set_option('display.max_colwidth', None)
# pd.set_option('display.width', None)

In [3]:
PROJECT_DIR = r"/kaggle/input/d/hsingh956/titanic/"


In [4]:
def get_data(file_name):
    file_name = f"{file_name}.csv"
    file_path = os.path.join(PROJECT_DIR, file_name)
    return pd.read_csv(file_path)

In [5]:
train = get_data("train")
train

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

In [6]:
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [7]:
train.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

## - data setcontains 891 entries and 11 feature columns 
## - Age and Cabin have missing values 177 and 687 each 


# Preliminary Analysis

In [8]:

train.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

### check datatypes


In [9]:
train.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

### Check Duplicates


In [10]:
train.duplicated().sum()

0

# Detailed Analysis


In [11]:
train.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [12]:
train.PassengerId



0        1
1        2
2        3
3        4
4        5
      ... 
886    887
887    888
888    889
889    890
890    891
Name: PassengerId, Length: 891, dtype: int64

In [13]:
train.Survived

0      0
1      1
2      1
3      1
4      0
      ..
886    0
887    1
888    0
889    1
890    0
Name: Survived, Length: 891, dtype: int64

In [14]:
train.Survived.unique()


array([0, 1])

In [15]:
train.Survived.isna().sum()

0

## Pclass


In [16]:
train.Pclass.unique()

array([3, 1, 2])

## Name


In [17]:
train.Name.unique().size

train['Title'] = train.Name.apply(lambda name: name.split(',')[1].split('.')[0].strip())
train


train = train.assign(
    Title=lambda df_: df_['Name']
                      .str.split(',')
                      .str[1]
                      .str.split('.')
                      .str[0]
                      .str.strip()
)
train

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked Title  
0        0         A/5 21171   7.2500   NaN        S    Mr  
1        0          PC 17599  71.2833   C85        C   Mrs  
2        0  STON/O2. 3101282   7.9250   NaN        S  Miss  
3        0            113803  53.1000  C123        S   Mrs  
4        0            373450   8.0500   NaN        S    Mr  
..     ...               ...      ...   ...      ...   ...  
886      0            211536  13.0000   NaN        S   Rev  
887      0            112053  30.0000   B42        S  Miss  
888      2        W./C. 6607  23.4500   NaN        S  Miss  
889      0            111369  30.0000  C148        C    Mr  
890      0            370376   7.7500   NaN        Q    Mr  

[891 rows x 13 columns]

In [18]:


df= (
    train
    .Name 
    .str.split(",", expand=True)
    .set_axis(["lastname", "firstname"], axis=1)
    .assign(
        lastname = lambda df_: (
            df_
            .lastname
            .str.strip()
        ),
        firstname= lambda df_: (
            df_
            .firstname
            .str.split("(")
            .str[0]
            .str.strip()
            . str.replace(r'(Mr\.|Mrs\.|Miss\.|Master\.)', '', regex=True)
            .str.split('\"')
            .str[0]
            .str.strip()
        )
    )
)


## Sex

In [19]:
train.Sex.unique()

array(['male', 'female'], dtype=object)

In [20]:
def is_male(ser):
    temp = ser.to_frame()
    return (
        temp
        .assign(is_male=np.select(
            [temp.iloc[:, 0] == "male", temp.iloc[:, 0] == "female"],
            [1, 0],
            default=-1  
        ))
    )


df = (
    train
    .Sex
    .pipe(is_male)
    .drop('Sex', axis=1)
    
)

## Age


In [21]:
train.Age.isna().sum()

train.Age[(train.Age <=0 )| (train.Age >=100)]

Series([], Name: Age, dtype: float64)

## SibSp


In [22]:
train.SibSp.unique()



array([1, 0, 3, 4, 2, 5, 8])

## Parch


In [23]:
train.Parch.unique()



array([0, 1, 2, 5, 3, 4, 6])

## Ticket


In [24]:
train.Ticket

0             A/5 21171
1              PC 17599
2      STON/O2. 3101282
3                113803
4                373450
             ...       
886              211536
887              112053
888          W./C. 6607
889              111369
890              370376
Name: Ticket, Length: 891, dtype: object

In [25]:
(
    train
    .assign(
        Ticket=lambda df_: (
            df_
            .Ticket
            .str.replace(' ' , "") 
        ),
        TicketPrefix = lambda df_: (
            df_ 
            .Ticket
            .str.extract(r'([A-Za-z\/\.]+)',expand=False).str.strip()
            .fillna('O')
        ),
        TicketNumber = lambda df_ : (
            df_
            .Ticket
            .str.extract(r'(\d+)', expand=False)
            .astype(float)
            .fillna(0)
        )
            
    )

)


PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch           Ticket     Fare Cabin Embarked Title TicketPrefix  \
0        0         A/521171   7.2500   NaN        S    Mr           A/   
1        0          PC17599  71.2833   C85        C   Mrs           PC   
2        0  STON/O2.3101282   7.9250   NaN        S  Miss       STON/O   
3        0           113803  53.1000  C123        S   Mrs            O   
4        0           373450   8.0500   NaN        S    Mr            O   
..     ...              ...      ...   ...      ...   ...          ...   
886      0           211536  13.0000   NaN        S   Rev            O   
887      0           112053  30.0000   B42        S  Miss            O   
888      2        W./C.6607  23.4500   NaN        S  Miss        W./C.   
889      0           111369  30.0000  C148        C    Mr            O   
890      0           370376   7.7500   NaN        Q    Mr            O   

     TicketNumber  
0        521171.0  
1         17599.0  
2             2.0  
3        113803.0  
4        373450.0  
..            ...  
886      211536.0  
887      112053.0  
888        6607.0  
889      111369.0  
890      370376.0  

[891 rows x 15 columns]

## Fare

In [26]:
train.Fare

0       7.2500
1      71.2833
2       7.9250
3      53.1000
4       8.0500
        ...   
886    13.0000
887    30.0000
888    23.4500
889    30.0000
890     7.7500
Name: Fare, Length: 891, dtype: float64

Cabin


In [27]:
train.Cabin.unique()

array([nan, 'C85', 'C123', 'E46', 'G6', 'C103', 'D56', 'A6',
       'C23 C25 C27', 'B78', 'D33', 'B30', 'C52', 'B28', 'C83', 'F33',
       'F G73', 'E31', 'A5', 'D10 D12', 'D26', 'C110', 'B58 B60', 'E101',
       'F E69', 'D47', 'B86', 'F2', 'C2', 'E33', 'B19', 'A7', 'C49', 'F4',
       'A32', 'B4', 'B80', 'A31', 'D36', 'D15', 'C93', 'C78', 'D35',
       'C87', 'B77', 'E67', 'B94', 'C125', 'C99', 'C118', 'D7', 'A19',
       'B49', 'D', 'C22 C26', 'C106', 'C65', 'E36', 'C54',
       'B57 B59 B63 B66', 'C7', 'E34', 'C32', 'B18', 'C124', 'C91', 'E40',
       'T', 'C128', 'D37', 'B35', 'E50', 'C82', 'B96 B98', 'E10', 'E44',
       'A34', 'C104', 'C111', 'C92', 'E38', 'D21', 'E12', 'E63', 'A14',
       'B37', 'C30', 'D20', 'B79', 'E25', 'D46', 'B73', 'C95', 'B38',
       'B39', 'B22', 'C86', 'C70', 'A16', 'C101', 'C68', 'A10', 'E68',
       'B41', 'A20', 'D19', 'D50', 'D9', 'A23', 'B50', 'A26', 'D48',
       'E58', 'C126', 'B71', 'B51 B53 B55', 'D49', 'B5', 'B20', 'F G63',
       'C62 C64',

In [28]:
train.Embarked.fillna('O').unique()

array(['S', 'C', 'Q', 'O'], dtype=object)

In [29]:
train.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
Title            0
dtype: int64

# Conclusion

### - need to change the index to PassangerID
### - Survived Column looks good 
### - Pclass column looks also good
### - Name column are changed to firstname and lastname
### - Sex change to is_male that is integer column
### - Age column looks good
### - SibSp column looks good
### - Parch column looks good
### - Extract two rows from Ticket => TicketPrefix and TicketNumber
### - Fare columns looks good 
### - Cabin also looks good 
### - Embark also looks good 



 ## Need to fill the values in the Age and the Cabin
 

Cleaning Function


In [30]:
def is_male(ser):
    temp = ser.to_frame()
    return (
        temp
        .assign(is_male=np.select(
            [temp.iloc[:, 0] == "male", temp.iloc[:, 0] == "female"],
            [1, 0],
            default=-1  
        ))
    )



def clean_data(df):
    return (
        df
        .drop_duplicates()
        .assign(
            Title= lambda df_ : (
                        df_
                        .Name
                      .str.split(',')
                      .str[1]
                      .str.split('.')
                      .str[0]
                      .str.strip()
            ),
            lastname=lambda df_: (
                df_
                .Name
                .str.split(",")
                .str[0]
                .str.strip()
            ),
            firstname=lambda df_: (
                df_
                .Name
                .str.split(",")
                .str[1]
                .str.split("(")
                .str[0]
                .str.replace(r'(Mr\.|Mrs\.|Miss\.|Master\.)', '', regex=True)
                .str.strip()
            ),
             Ticket=lambda df_: (
            df_
            .Ticket
            .str.replace(' ' , "") 
            ),
            TicketPrefix = lambda df_: (
                df_ 
                .Ticket
                .str.extract(r'([A-Za-z\/\.]+)',expand=False).str.strip()
                .fillna('O')
            ),
            TicketNumber = lambda df_ : (
                df_
                .Ticket
                .str.extract(r'(\d+)', expand=False)
                .astype(float)
                .fillna(0)
            ),     
        )
        .drop('Name', axis=1)
        .set_index('PassengerId')
    )
cleaned_train = clean_data(train)

In [31]:
test = get_data("test")

cleaned_test = clean_data(test)
cleaned_test

Pclass     Sex   Age  SibSp  Parch             Ticket      Fare  \
PassengerId                                                                    
892               3    male  34.5      0      0             330911    7.8292   
893               3  female  47.0      1      0             363272    7.0000   
894               2    male  62.0      0      0             240276    9.6875   
895               3    male  27.0      0      0             315154    8.6625   
896               3  female  22.0      1      1            3101298   12.2875   
...             ...     ...   ...    ...    ...                ...       ...   
1305              3    male   NaN      0      0           A.5.3236    8.0500   
1306              1  female  39.0      0      0            PC17758  108.9000   
1307              3    male  38.5      0      0  SOTON/O.Q.3101262    7.2500   
1308              3    male   NaN      0      0             359309    8.0500   
1309              3    male   NaN      1      1               2668   22.3583   

            Cabin Embarked   Title       lastname        firstname  \
PassengerId                                                          
892           NaN        Q      Mr          Kelly            James   
893           NaN        S     Mrs         Wilkes            James   
894           NaN        Q      Mr          Myles   Thomas Francis   
895           NaN        S      Mr           Wirz           Albert   
896           NaN        S     Mrs       Hirvonen        Alexander   
...           ...      ...     ...            ...              ...   
1305          NaN        S      Mr        Spector            Woolf   
1306         C105        C    Dona  Oliva y Ocana    Dona. Fermina   
1307          NaN        S      Mr        Saether  Simon Sivertsen   
1308          NaN        S      Mr           Ware        Frederick   
1309          NaN        C  Master          Peter        Michael J   

            TicketPrefix  TicketNumber  
PassengerId                             
892                    O      330911.0  
893                    O      363272.0  
894                    O      240276.0  
895                    O      315154.0  
896                    O     3101298.0  
...                  ...           ...  
1305                  A.           5.0  
1306                  PC       17758.0  
1307          SOTON/O.Q.     3101262.0  
1308                   O      359309.0  
1309                   O        2668.0  

[418 rows x 14 columns]

In [32]:
cleaned_train
cleaned_test

Pclass     Sex   Age  SibSp  Parch             Ticket      Fare  \
PassengerId                                                                    
892               3    male  34.5      0      0             330911    7.8292   
893               3  female  47.0      1      0             363272    7.0000   
894               2    male  62.0      0      0             240276    9.6875   
895               3    male  27.0      0      0             315154    8.6625   
896               3  female  22.0      1      1            3101298   12.2875   
...             ...     ...   ...    ...    ...                ...       ...   
1305              3    male   NaN      0      0           A.5.3236    8.0500   
1306              1  female  39.0      0      0            PC17758  108.9000   
1307              3    male  38.5      0      0  SOTON/O.Q.3101262    7.2500   
1308              3    male   NaN      0      0             359309    8.0500   
1309              3    male   NaN      1      1               2668   22.3583   

            Cabin Embarked   Title       lastname        firstname  \
PassengerId                                                          
892           NaN        Q      Mr          Kelly            James   
893           NaN        S     Mrs         Wilkes            James   
894           NaN        Q      Mr          Myles   Thomas Francis   
895           NaN        S      Mr           Wirz           Albert   
896           NaN        S     Mrs       Hirvonen        Alexander   
...           ...      ...     ...            ...              ...   
1305          NaN        S      Mr        Spector            Woolf   
1306         C105        C    Dona  Oliva y Ocana    Dona. Fermina   
1307          NaN        S      Mr        Saether  Simon Sivertsen   
1308          NaN        S      Mr           Ware        Frederick   
1309          NaN        C  Master          Peter        Michael J   

            TicketPrefix  TicketNumber  
PassengerId                             
892                    O      330911.0  
893                    O      363272.0  
894                    O      240276.0  
895                    O      315154.0  
896                    O     3101298.0  
...                  ...           ...  
1305                  A.           5.0  
1306                  PC       17758.0  
1307          SOTON/O.Q.     3101262.0  
1308                   O      359309.0  
1309                   O        2668.0  

[418 rows x 14 columns]